pkg/clients/kube/kube.go

```diff
@@ -0,0 +1,149 @@
+// Generated by Claude 3.5 Sonnet
+package kube
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/kubernetes"
+)
+
+// Client represents a Kubernetes Client
+type Client struct {
+	clientset kubernetes.Interface
+	namespace string
+}
+
+// NewClient creates a new Kubernetes client from a clientset
+func NewClient(clientset *kubernetes.Clientset) (*Client, error) {
+	return &Client{
+		clientset: clientset,
+		namespace: "default",
+	}, nil
+}
+
+// SetNamespace sets the namespace for the client
+func (c *Client) SetNamespace(namespace string) {
+	c.namespace = namespace
+}
+
+// CreateJob creates a Kubernetes Job with the provided spec
+func (c *Client) CreateJob(ctx context.Context, job *batchv1.Job) (*batchv1.Job, error) {
+	// Sanity check
+	if job.Name == "" {
+		return nil, fmt.Errorf("job name is empty")
+	}
+
+	return c.clientset.BatchV1().Jobs(c.namespace).Create(ctx, job, metav1.CreateOptions{})
+}
+
+// GetJob retrieves a Kubernetes Job by name
+func (c *Client) GetJob(ctx context.Context, jobName string) (*batchv1.Job, error) {
+	return c.clientset.BatchV1().Jobs(c.namespace).Get(ctx, jobName, metav1.GetOptions{})
+}
+
+// DeleteJob deletes a Kubernetes Job by name
+func (c *Client) DeleteJob(ctx context.Context, jobName string) error {
+	propagationPolicy := metav1.DeletePropagationForeground
+	return c.clientset.BatchV1().Jobs(c.namespace).Delete(ctx, jobName, metav1.DeleteOptions{PropagationPolicy: &propagationPolicy})
+}
+
+// WatchJob watches a Kubernetes Job for changes
+func (c *Client) WatchJob(ctx context.Context, jobName string) (watch.Interface, error) {
+	return c.clientset.BatchV1().Jobs(c.namespace).Watch(ctx, metav1.ListOptions{
+		FieldSelector: fmt.Sprintf("metadata.name=%s", jobName),
+	})
+}
+
+// GetJobLogs retrieves logs from a Kubernetes Job's pods
+func (c *Client) GetJobLogs(ctx context.Context, jobName string) (string, error) {
+	// Get the pods for this job
+	pods, err := c.clientset.CoreV1().Pods(c.namespace).List(ctx, metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("job-name=%s", jobName),
+	})
+	if err != nil {
+		return "", fmt.Errorf("failed to list pods for job %s: %w", jobName, err)
+	}
+
+	if len(pods.Items) == 0 {
+		return "", fmt.Errorf("no pods found for job %s", jobName)
+	}
+
+	// Get logs from the first pod (assuming single pod job)
+	pod := pods.Items[0]
+	logs, err := c.clientset.CoreV1().Pods(c.namespace).GetLogs(pod.Name, &corev1.PodLogOptions{}).Do(ctx).Raw()
+	if err != nil {
+		return "", fmt.Errorf("failed to get logs for pod %s: %w", pod.Name, err)
+	}
+
+	return string(logs), nil
+}
+
+// WaitForJobCompletion waits for a job to complete (succeed or fail)
+func (c *Client) WaitForJobCompletion(ctx context.Context, jobName string, timeout time.Duration) error {
```

Comment:
Do we need to pass in both a context and a timeout or can the context that's passed in handle it? We don't have a way to cancel the WatchJob function since the passed in context is just `TODO`.

Reply:
I have usually seen the following:
```
context.WithTimeout(context.Background(), time.Second*3)
```

Reply:
Oh I see you are doing this on line 96 :)

Reply:
We discussed this in handover last week, and the outcome was that we'll just enforce a timeout via a context that's configured by the caller instead of having a dedicated timeout parameter